public/app/plugins/panel/singlestat2/ProcessedValuesRepeater.tsx

```diff
@@ -0,0 +1,48 @@
+import React, { PureComponent } from 'react';
+import { VizOrientation } from '@grafana/ui';
+import { VizRepeater } from '@grafana/ui';
+
+export interface Props<T> {
+  width: number;
+  height: number;
+  orientation: VizOrientation;
+  source: any; // If this changes, the values will be processed
+  processFlag?: boolean; // change to force processing
+
+  getProcessedValues: () => T[];
+  renderValue: (value: T, width: number, height: number) => JSX.Element;
+}
+
+interface State<T> {
+  values: T[];
+}
+
+/**
+ * This is essentially a cache of processed values.  This checks for changes
+ * to the source and then saves the processed values in the State
+ */
+export class ProcessedValuesRepeater<T> extends PureComponent<Props<T>, State<T>> {
```

Comment:
wonder if would make sense to merge this with VizRepeater (not sure)

Reply:
I had that at first, but I wanted to think about what moves to grafana/ui in a different PR.  This is already kind of a monster :)

But yes, it is a variation of VizRepeater that holds processed values in State.  Maybe VizRepeater -> Repeater, and ProcessedValuesRepeater -> RepeaterWithCache?  but lets discuss that in another issue